pkg/security/admission/admission.go

```diff
@@ -0,0 +1,318 @@
+package admission
+
+import (
+	"fmt"
+	"io"
+
+	kadmission "github.com/GoogleCloudPlatform/kubernetes/pkg/admission"
+	kapi "github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api/errors"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client/cache"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/fields"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/runtime"
+	scc "github.com/GoogleCloudPlatform/kubernetes/pkg/securitycontextconstraints"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/serviceaccount"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/watch"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/auth/user"
+	allocator "github.com/openshift/origin/pkg/security"
+	"github.com/openshift/origin/pkg/security/uid"
+)
+
+func init() {
+	kadmission.RegisterPlugin("SecurityContextConstraint", func(client client.Interface, config io.Reader) (kadmission.Interface, error) {
+		return NewConstraint(client), nil
+	})
+}
+
+type constraint struct {
+	*kadmission.Handler
+	client client.Interface
+	store  cache.Store
+}
+
+var _ kadmission.Interface = &constraint{}
+
+func NewConstraint(kclient client.Interface) kadmission.Interface {
+	store := cache.NewStore(cache.MetaNamespaceKeyFunc)
+	reflector := cache.NewReflector(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return kclient.SecurityContextConstraints().List(labels.Everything(), fields.Everything())
+			},
+			WatchFunc: func(resourceVersion string) (watch.Interface, error) {
+				return kclient.SecurityContextConstraints().Watch(labels.Everything(), fields.Everything(), resourceVersion)
+			},
+		},
+		&kapi.SecurityContextConstraints{},
+		store,
+		0,
+	)
+	reflector.Run()
+
+	return &constraint{
+		Handler: kadmission.NewHandler(kadmission.Create, kadmission.Update),
+		client:  kclient,
+		store:   store,
+	}
+}
+
+// Admit determines if the pod should be admitted based on the requested security context
+// and the available SCCs.
+//
+// 1.  Find all the SCCs that the service account or user has access to
+// 2.  Full resolve each SCC
+//     1.  Determine if the user/selinux strategies requires pre-allocated values
+//     2.  Retrieve pre-allocated values from the namespace and set them on the strategies
+// 3.  Create the set of providers from the fully configured SCCs
+// 4.  Generate and validate each container's SC against the set of SCC providers.  If any container
+//     fails to validate then the entire request is rejected
+func (c *constraint) Admit(a kadmission.Attributes) error {
+	if a.GetResource() != string(kapi.ResourcePods) {
+		return nil
+	}
+
+	pod, ok := a.GetObject().(*kapi.Pod)
+	// if we can't convert then we don't handle this object so just return
+	if !ok {
+		return nil
+	}
+
+	// get all constraints that are usable by the user
```

Comment:
quick overview of what I think this method should do:
1. figure out who we're checking (get user's name/groups, service account's name/groups)
2. find applicable SCCs, deduping ones that apply multiple times (to both user and service account, or to multiple groups, etc)
3. convert constraints to providers, resolving required inputs where needed (like looking up namespace annotations), collecting any errors. should only attempt to fetch the namespace once, even if multiple constraints/strategies would make use of the annotations
4. See if any provider allows the pod (collecting any errors while evaluating), and if so, return early. The current approach allows each container's security context to be determined by a different provider. That seems odd to me... it seems like all the containers in the pod should be validated with a single provider, all or nothing. Is there a strong argument otherwise?
5. If no providers existed, or no provider allowed all the containers, return nil or an aggregated error if errors occurred.